firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java

```diff
@@ -329,14 +330,27 @@ public SnapshotVersion getLastRemoteSnapshotVersion() {
             }
           }
 
-          Set<DocumentKey> changedDocKeys = new HashSet<>();
+          Map<DocumentKey, MaybeDocument> changedDocs = new HashMap<>();
           Map<DocumentKey, MaybeDocument> documentUpdates = remoteEvent.getDocumentUpdates();
           Set<DocumentKey> limboDocuments = remoteEvent.getResolvedLimboDocuments();
+
+          Set<DocumentKey> keys = new HashSet<>();
+          for (Entry<DocumentKey, MaybeDocument> entry : documentUpdates.entrySet()) {
+            keys.add(entry.getKey());
+          }
+          List<MaybeDocument> existingDocs = remoteDocuments.getAll(keys);
```

Comment:
Optimization 2 ("No get by one"): get all the documents from the local database in a single query.

IIUC, keys are not repeated, so retrieving all the documents before going into the main for loop is okay, because no iteration of the loop may affect subsequent iterations.